hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java

```diff
@@ -354,6 +354,15 @@ ContainerCommandResponseProto handleCreateContainer(
     }
 
     long containerID = request.getContainerID();
+    State containerState = request.getCreateContainer().getState();
+
+    if (containerState != RECOVERING) {
+      try {
+        containerSet.validateContainerIsMissing(containerID, containerState);
+      } catch (StorageContainerException ex) {
+        return ContainerUtils.logAndReturnError(LOG, ex, request);
+      }
+    }
```

Comment:
This code reads poorly and complicates a straighforward logic. The witnessed container list should NOT care about the state. 
```suggestion
      if (containerSet.wasContainerWitnessed(containerID) && containerState != RECOVERING) {
         LOG.warn("Trying to create a container already witnessed on this node which is not in recovering state);
         return ContainerUtils.logAndReturnError(LOG, null, request);
      }
      
```

Reply:
Had a discussion offline with @kerneltime . We have to perform this state check because EC container replication path follows the dispatcher flow. In order to allow replication manager to re create a missing container on the same node, we have to perform this check and override the check.